Build a URL Shortener with React, TypeScript, and Directus (#223)
directus-labs/guest-authoring: Build a URL Shortener with React, TypeScript, and Directus/index.md.md

We will be using _[vite](https://vitejs.dev/)_ to create our react app.

Install React.js by running the following commands.

```
 npm create vite@latest link-shortener

  # Select React as the framework and Typescript as a variant

 npm install
```

Also, the `types/node` package.

```
npm i --save-dev @types/node
```

Run the scaffolded React.js application with this command

```
 npm run dev
```

Now, open http://localhost:5173/ on your browser, and you should see the Vite(react) starter page.

![Vite_React](<React(Vite) Project Setup with Typescript.PNG>)

Next, we will set up Vite to allow environment variables, which will be used to store our Directus credentials.

## Directus Tokens and Environment Variables

In the _vite.config.ts_ file, add the following code:

```typescript
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  define: {
    "process.env.VITE_DIRECTUS_API_TOKEN.": JSON.stringify(
      process.env.VITE_DIRECTUS_API_TOKEN
    ),
    "process.env.VITE_DIRECTUS_API_URL.": JSON.stringify(
      process.env.VITE_DIRECTUS_API_URL
    ),
  },

> });
> ```
> 
> Create a _.env_ file in the _src_ directory, and the following code:
> 
> ```
> VITE_DIRECTUS_API_TOKEN = XXXXX
> VITE__DIRECTUS_API_URL = https://linksshortener.directus.app/
> ```
> 
> To create a Static Token for [authentication](https://docs.directus.io/reference/authentication.html) do the following:
> 
> 1. On the Directus dashboard, navigate to **User Directory > Administrator**
> 2. Click on the **Admin User** icon.
> 3. Scroll down to the **Admin Options** section
> 4. Click on the + icon in the **Token** field to generate token.
> 5. Copy the generated token and paste it in `VITE_DIRECTUS_API_TOKEN`.
> 
> ![Admin_Token](<Directus Tokens and Environment Variables_1.PNG>)
> 
> For the project URL, copy the URL from your browser.
> 
> ![App_URL](<Directus Tokens and Environment Variables_2.PNG>)
> 
> ## Integrating the Directus SDK with React
> 
> Install the SDK using the command:
> 
> ```
> npm install @directus/sdk
> ```
> 
> ### Import the SDK Composables
> 
> In the _src_ directory create a sub-directory called _utils_, in _utils_ create a _directus.ts_ file.
> 
> Add the following code to _directus.ts_:
> 
> ```typescript
> import { createDirectus, staticToken, rest } from "@directus/sdk";
> 
> const directusToken = import.meta.env.VITE_DIRECTUS_API_TOKEN;
> const directusUrl = import.meta.env.VITE__DIRECTUS_API_URL;
> 
> if (!directusToken) {
>   throw new Error("Please include a Token");
> }
> 
> if (!directusUrl) {
>   throw new Error("Please include a Url");
> }
> 
> export const directus = createDirectus(directusUrl)
>   .with(staticToken(directusToken))
>   .with(rest());
> ```
> 
> - `staticToken`: is used to authenticate the client we'll create.
> - `createDirectus`: is a function that initializes a Directus client.
> - `rest()`: enables the `.request(...)` method to query the collection.

Reviewer: ````suggestion
## Setting Up Your React Project

Install React.js, set up dependencies, and run a development server:

```
npm create vite@latest link-shortener
? Select a framework: React
? Select a variant: TypeScript

npm install
npm install --save-dev @types/node
npm install react-router-dom
npm install @directus/sdk 

npm run dev
```

Now, open http://localhost:5173/ on your browser, and you should see the starter page.

Next, we will set up Vite to allow environment variables, which will be used to store our Directus credentials. In the `vite.config.ts` file, add the following code:

```typescript
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";

export default defineConfig({
  plugins: [react()],
  define: {
    "process.env.VITE_DIRECTUS_API_TOKEN.": JSON.stringify(
      process.env.VITE_DIRECTUS_API_TOKEN
    ),
    "process.env.VITE_DIRECTUS_API_URL.": JSON.stringify(
      process.env.VITE_DIRECTUS_API_URL
    ),
  },
});
```

Create a `.env` file in the `src` directory, and the following variables, being sure to provide your specific static authentication token and project URL:

```
VITE_DIRECTUS_API_TOKEN = XXXXX
VITE__DIRECTUS_API_URL = https://your-amazing.directus.app/
```

In the `src` directory create a sub-directory called `utils`, and within it, a `directus.ts` file.

Add the following code to _directus.ts_:

```typescript
import { createDirectus, staticToken, rest } from "@directus/sdk";

const directusToken = import.meta.env.VITE_DIRECTUS_API_TOKEN;
const directusUrl = import.meta.env.VITE__DIRECTUS_API_URL;

if (!directusToken) {
  throw new Error("Please include a Token");
}

if (!directusUrl) {
  throw new Error("Please include a Url");
}

export const directus = createDirectus(directusUrl)
  .with(staticToken(directusToken))
  .with(rest());
```
````

> ### Creating the Dynamic Route
> 
> Let's create the route that will be used to redirect the user to the slug URL queried from Directus.
> 
> The `react-router-dom` library will be used to handle client-side routing, install the `react-router-dom` library by running the command:
> 
> ```
> npm install react-router-dom
> ```
> 
> Next, in the _App.tsx_ file add the following code:
> 
> ```typescript
> import { BrowserRouter, Routes, Route } from "react-router-dom";
> import { useParams } from "react-router-dom";
> 
> function App() {
>   return (
>     <>
>       <h1>Link Shortener</h1>
>       <BrowserRouter>
>         <Routes>
>           <Route path="/:slug" element={<LinkRoute />}></Route>
>         </Routes>
>       </BrowserRouter>
>     </>
>   );
> }
> 
> function LinkRoute() {
>   const { slug } = useParams();
> 
>   console.log(slug);
> 
>   return (
>     <>
>       <h1>{slug}</h1>
>     </>
>   );
> }
> 
> export default App;
> ```
> 
> Browser Output:
> 
> ![Dynamic URL Output](<Creating the Dynamic Route (1).gif>)

Reviewer: ````suggestion
### Creating the Dynamic Route

Let's create the route that will be used to redirect the user to the slug URL queried from Directus. In the `App.tsx` file add the following code:

```typescript
import { BrowserRouter, Routes, Route } from "react-router-dom";
import { useParams } from "react-router-dom";

function App() {
  return (
    <>
      <h1>Link Shortener</h1>
      <BrowserRouter>
        <Routes>
          <Route path="/:slug" element={<LinkRoute />}></Route>
        </Routes>
      </BrowserRouter>
    </>
  );
}

function LinkRoute() {
  const { slug } = useParams();

  console.log(slug);

  return (
    <>
      <h1>{slug}</h1>
    </>
  );
}

export default App;
```
````

The `LinkRoute` component uses the `useParams` hook to get the slug parameter from the URL. It also uses the `useState` hook to store an error message if the slug is invalid.

The `useEffect` hook is used to fetch data from the Directus API via the `fetchShortLink` effect function, which performs the following actions:

> The `ShortLink` interface defines the shape of the short link data returned by the Directus API.

1. It makes a request to the Directus API to read all short links via the `readItems` composable.
2. The response data is filtered to find the short link that matches the current slug.
3. If no matching short link is found, it sets an error message and throws an error.
4. If a matching short link is found, it updates the short links `click` property by `1` via the `updateItem` composable from Directus API.
5. Finally, the user is redirected to the URL associated with the short link using the `window.location.assign` method.

Finally, if an error occurs during the data fetching process, the error is caught and logged to the console. The `LinkRoute` component also returns the error message if the slug is invalid.

Browser Output:

> ![Slug URL Output](<Querying Directus.gif>)

Reviewer: Please can you take this section into smaller, more digestible steps. Please also make use of our diff formatting (check the writing guide) to draw attention to additions/deletions.

Author: Done.

> ## Conclusion
> 
> You’ve just learned how to set up the data collection process and implement dynamic routing using React and Directus. The broad process was:
> 
> - Set up Directus data model
> - Customize user roles and permissions
> - API endpoint to query data from Directus
> - Dynamic Route to navigate to slug URL
> 
> I hope you find this tutorial useful - if you have any questions or hurdles feel free to join the Directus [official Discord server](https://discord.com/invite/directus).

Reviewer: ```suggestion
## Conclusion

You’ve just learned how to set up the data collection process and implement dynamic routing using React and Directus. The broad process involved setting up a Directus data model, customizing user roles and permissions, creating an API endpoint to query data from Directus, and configuring a dynamic route to navigate to slug URLs.

I hope you find this tutorial useful - if you have any questions or hurdles feel free to join the Directus [official Discord server](https://directus.chat).
```
